include/tlapack/lapack/gelqf.hpp

```diff
@@ -128,29 +136,41 @@ int gelqf(matrix_t& A,
     const idx_t nb = opts.nb;
 
     // check arguments
-    tlapack_check_false(nrows(TT) < m || ncols(TT) < nb);
+    tlapack_check((idx_t)size(tau) >= k);
 
-    // Allocates workspace
+    // Allocate or get workspace
+    Workspace sparework;
     vectorOfBytes localworkdata;
     Workspace work = [&]() {
         workinfo_t workinfo;
-        gelqf_worksize(A, TT, workinfo, opts);
+        gelqf_worksize(A, tau, workinfo, opts);
         return alloc_workspace(localworkdata, workinfo, opts.work);
     }();
 
-    // Options to forward
-    auto&& gelq2Opts = workspace_opts_t<>{work};
+    // TT serves two functions, it stores the triangular factors
+    // in the compact WY representation and the parts that do not
+    // store triangular factors yet are used as workspace.
+    // If passed as an optional arguments, it is used so that the
+    // triangular factors can be reused later. Otherwise, workspace is used.
+    matrix_t TT = [&]() {
```

Comment:
I think the problem is here. Try replacing `matrix_t` by `auto`. I explain:

- When `*opts.TT` is returned, the type of matrix `TT` is `matrix_t`. We're fine.
- When `new_matrix(work, m, nb, sparework)` is returned, we have whatever `new_matrix` returns. In case of Eigen, `new_matrix` returns a `Eigen::Map<matrix_t>`. This is the class in Eigen that maps pointers to raw data to a Eigen matrix.

It is listed in [CONTRIBUTING.md](https://github.com/tlapack/tlapack/blob/master/CONTRIBUTING.md) at "We recommend the usage of auto in the following cases:".

Reply:
I had the same idea, but the problem is slightly earlier. The compiler needs to know the return type of the lambda function. In the situation outlined in contributing.md, the return type is determined at compile time.

Reply:
having a top level routine might solve it, then it can be encapsulated in two instantiations of the same function